conans/client/graph/proxy.py

```diff
@@ -39,30 +39,43 @@ def _get_recipe(self, reference, check_updates, update, remote_name):
             status = RECIPE_DOWNLOADED
             return conanfile_path, status, remote, new_ref
 
+        # In disk but not with the same revision
+        if reference.revision:
+            full_ref = self._registry.get_ref_with_revision(reference)
+            if full_ref.revision != reference.revision:
+                output.info("Requested different revision of %s..." % str(reference))
+                remote, new_ref = self._download_recipe(reference, output, remote_name)
+                status = RECIPE_UPDATED
+                return conanfile_path, status, remote, new_ref
+
         remote = self._registry.get_recipe_remote(reference)
+
         check_updates = check_updates or update
         # Recipe exists in disk, but no need to check updates
         if not check_updates:
             status = RECIPE_INCACHE
             log_recipe_got_from_local_cache(reference)
             self._recorder.recipe_fetched_from_cache(reference)
-            return conanfile_path, status, remote, reference
+            new_ref = self._registry.get_ref_with_revision(reference) or reference
```

Comment:
Same line ``self._registry.get_ref_with_revision`` is done in many places. Maybe would be better to do it just once in the caller of ``ConanProxy.get_recipe()``

Reply:
I think you cannot. Because the registry is modified when intermediate downloads are being executed.

Reply:
But it would be the same result, wouldnt it?, you get the return from calling get recipe and compute the new_ref calling ``self._registry.get_ref_with_revision()``. I see it just as a refactor of code, not changing functionality.

Reply:
ah! I think now I get it, let me check.